_plugins/pa11y.rb

```diff
@@ -0,0 +1,177 @@
+=begin
+
+Determines which files to check with pa11y-ci per pull request.
+
+The current scope of pa11y testing is as follows:
+
+- Check any posts or pages whose content has changed.
+- Check any posts or pages using changed layouts.
+- If stylesheets have changed, check a sample of all page types:
+    - Three posts of each layout in _config.yml's `defaults`
+    - The first, last, and middle blog archive pages (/blog/page/:num)
+
+Previously, checking every page resulted in excessive CI runtimes.
+
+=end
+
+# Wraps Jekyll documents with convenience methods.
+#
+# @param path [String] The relative path of the source file, usually a
+#    Markdown file (.md)
+# @param layout [String] The name of the Jekyll document's layout
+# @param differ [Object, #changed_files] The collaborator which lists
+#   the files that were changed in the last commit.
+# @todo Are there cases when the document being checked has no layout?
+Document = Struct.new(:path, :layout, :differ) do
+  # @return [Boolean] Whether to scan this document in pa11y-ci
+  def to_scan?
+    source_changed? || layout_changed?
+  end
+
+  def source_changed?
+    changed_files.include?(path)
+  end
+
+  def layout_changed?
+    changed_files.include?("_layouts/#{layout}.html")
+  end
+
+  # Memoize these without having to memoize in the differ
+  def changed_files
+    @changed_files ||= differ.changed_files
+  end
+end
+
+class Differ
+  # @return [Array<String>] List of relative paths of files
+  def self.changed_files
+    @changed_files ||= %x[ #{list_files_command} ].to_s.split("\n").map(&:strip)
+  end
+
+  # @return [String] Text of a shell command to get listed files
+  def self.list_files_command
+    "git ls-files --modified"
+  end
+end
+
+# Gets the list of files changed in the last commit
+class CommitDiffer < Differ
+  def self.list_files_command
+    "git diff-tree --no-commit-id --name-only -r $(git rev-parse HEAD)"
+  end
+end
+
+# Jekyll::Document#destination requires a root path parameter
+# This constant replaces a mysterious empty string when
+# calling #destination.
+NO_ROOT_PATH = "".freeze
+PA11Y_TARGET_FILE = ENV.fetch('PA11Y_TARGET_FILE') { 'pa11y_targets' }
+DIFFER = ENV.fetch("CI", false) ? CommitDiffer : Differ
+
+def check_file_for_changes(file)
```

Comment:
suggestion: This function is at the top level, so when it's `required` into the test file, we'll have a free-floating function. **I'd suggest, at minimum, organizing this into a module,** probably a new one.

suggestions (non-blocking): My dream is to have this function be tested, and that would be easier if the pa11y target file and differ constants were both able to be dependency-injected.